src/auth/src/credentials/external_account_sources/executable_sourced.rs

```diff
@@ -0,0 +1,317 @@
+                return Err(CredentialsError::from_msg(true, msg));
+            };
+            return Err(CredentialsError::from_msg(true, msg));
+        }
+
+        let subject_token = String::from_utf8(output.stdout)
+            .map_err(|e| CredentialsError::from_source(true, e))?
+            .to_string();
+
+        Ok(subject_token)
+    }
+
+    fn split_command(command: String) -> (String, Vec<String>) {
+        let mut parts = command.split_whitespace();
+
+        if let Some(command) = parts.next() {
+            let args: Vec<String> = parts.map(String::from).collect();
+            return (command.to_string(), args);
+        }
+
+        (command, vec![])
+    }
+
+    fn parse_token(output: String) -> Result<String> {
+        let res = serde_json::from_str::<ExecutableResponse>(output.as_str())
+            .map_err(|e| CredentialsError::from_source(false, e))?;
+
+        if !res.success {
+            return Err(res.to_cred_error());
+        }
+
+        match res.token_type.as_str() {
+            JWT_TOKEN_TYPE | ACCESS_TOKEN_TYPE => match res.id_token {
+                Some(id_token) => Ok(id_token),
+                None => Err(CredentialsError::from_msg(
+                    false,
+                    "missing `id_token` field",
+                )),
+            },
+            SAML2_TOKEN_TYPE => match res.saml_response {
+                Some(saml_response) => Ok(saml_response),
+                None => Err(CredentialsError::from_msg(
+                    false,
+                    "missing `saml_response` field",
+                )),
+            },
+            _ => Err(CredentialsError::from_msg(
+                false,
+                "contains unsupported token type",
+            )),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::constants::JWT_TOKEN_TYPE;
+    use scoped_env::ScopedEnv;
+    use serde_json::json;
+    use serial_test::serial;
+    use tokio::time::{Duration, Instant};
+
+    type TestResult = std::result::Result<(), Box<dyn std::error::Error>>;
+
+    #[tokio::test]
+    #[serial]
+    async fn read_token_from_command() -> TestResult {
+        let _e = ScopedEnv::set(ALLOW_EXECUTABLE_ENV, "1");
+        let expiration = (Instant::now() + Duration::from_secs(3600))
+            .elapsed()
+            .as_millis();
+        let json_response = json!({
+            "success": true,
+            "version": 1,
+            "expiration_time": expiration,
+            "token_type": JWT_TOKEN_TYPE,
+            "id_token":"an_example_token",
+        })
+        .to_string();
+        let file = tempfile::NamedTempFile::new().unwrap();
+        let path = file.into_temp_path();
+        std::fs::write(&path, json_response).expect("Unable to write to temp file with command");
+
+        let token_provider = ExecutableSourcedCredentials {
+            executable: ExecutableConfig {
+                command: Some(format!("cat {}", path.to_str().unwrap())),
```

Comment:
How do you know that `cat` is in the path?  That may be true in the CI build for Windows, but won't be true in general.

Also, how do you know that `file` does not have a space?

Reply:
that's a good question, I was having a lot of trouble running these tests, specially the one with timeout, as I was trying to use the `sleep` command and it was not available on `macos` (tho it is available on my mac machine here). Do you have any recommendation on a better way to test this ? 

On the `file` name, AFAIK `temp` files don't have spaces on it.

Reply:
> that's a good question, I was having a lot of trouble running these tests, specially the one with timeout, as I was trying to use the `sleep` command and it was not available on `macos` (tho it is available on my mac machine here). Do you have any recommendation on a better way to test this ?

I believe you can create a binary target in `src/auth/${new target}`, take a dependency on this crate, and then run that binary from your program. You will need to compute the path, but there are variables to discover it.

> 
> On the `file` name, AFAIK `temp` files don't have spaces on it.

Until you run into somebody who sets `export TMPDIR="/tmp/some people like spaces in their  paths/foo"`.  And on Windows they are even more common.

Reply:
BTW, an assertion that says "this test does not work if temporary files have spaces" would be enough.